File: cdm-sources/src/cdm/margin/schedule/functions/GetStandardizedScheduleMarginRate.py
# pylint: disable=line-too-long, invalid-name, missing-function-docstring, missing-module-docstring, superfluous-parens
# pylint: disable=wrong-import-position, unused-import, unused-wildcard-import, wildcard-import, wrong-import-order, missing-class-docstring
from __future__ import annotations
import sys
import datetime
import inspect
from decimal import Decimal
from rosetta.runtime.utils import *
from rosetta.runtime.func_proxy import replaceable, create_module_attr_guardian
from cdm.margin.schedule.StandardizedScheduleAssetClassEnum import StandardizedScheduleAssetClassEnum

__all__ = ['GetStandardizedScheduleMarginRate']


@replaceable
def GetStandardizedScheduleMarginRate(assetClass: StandardizedScheduleAssetClassEnum, durationInYears: Decimal) -> Decimal:
    """
    Computes the margin rate, which is required in the calculation of the gross initial margin. It depends exclusively on the asset class of the trade and, in some cases, on the duration as well.
    
    Parameters 
    ----------
    assetClass : StandardizedScheduleAssetClassEnum
    
    durationInYears : number
    
    Returns
    -------
    percentage : number
    
    """
    self = inspect.currentframe()
    
    
    def _then_fn3():
        return 4.0
    
    def _else_fn3():
        return True
    
    def _then_fn2():
        return 2.0
    
    def _else_fn2():
        return if_cond_fn(all_elements(rosetta_resolve_attr(self, "durationInYears"), ">", 5), _then_fn3, _else_fn3)
    
    def _then_fn1():
        return 1.0
    
    def _else_fn1():
        return if_cond_fn((all_elements(rosetta_resolve_attr(self, "durationInYears"), ">", 2) and all_elements(rosetta_resolve_attr(self, "durationInYears"), "<=", 5)), _then_fn2, _else_fn2)
    
    def _then_fn4():
        return 10.0
    
    def _else_fn4():
        return True
    
    def _then_fn3():
        return 5.0
    
    def _else_fn3():
        return if_cond_fn(all_elements(rosetta_resolve_attr(self, "durationInYears"), ">", 5), _then_fn4, _else_fn4)
    
    def _then_fn2():
        return 2.0
    
    def _else_fn2():
        return if_cond_fn((all_elements(rosetta_resolve_attr(self, "durationInYears"), ">", 2) and all_elements(rosetta_resolve_attr(self, "durationInYears"), "<=", 5)), _then_fn3, _else_fn3)
    
    def _then_fn4():
        return 15.0
    
    def _else_fn4():
        return True
    
    def _then_fn3():
        return 15.0
    
    def _else_fn3():
        return if_cond_fn(all_elements(rosetta_resolve_attr(self, "assetClass"), "=", rosetta_resolve_attr(StandardizedScheduleAssetClassEnum, "COMMODITY")), _then_fn4, _else_fn4)
    
    def _then_fn2():
        return 6.0
    
    def _else_fn2():
        return if_cond_fn(all_elements(rosetta_resolve_attr(self, "assetClass"), "=", rosetta_resolve_attr(StandardizedScheduleAssetClassEnum, "EQUITY")), _then_fn3, _else_fn3)
    
    def _then_fn1():
        return if_cond_fn(all_elements(rosetta_resolve_attr(self, "durationInYears"), "<=", 2), _then_fn2, _else_fn2)
    
    def _else_fn1():
        return if_cond_fn(all_elements(rosetta_resolve_attr(self, "assetClass"), "=", rosetta_resolve_attr(StandardizedScheduleAssetClassEnum, "FOREIGN_EXCHANGE")), _then_fn2, _else_fn2)
    
    def _then_fn0():
        return if_cond_fn(all_elements(rosetta_resolve_attr(self, "durationInYears"), "<=", 2), _then_fn1, _else_fn1)
    
    def _else_fn0():
        return if_cond_fn(all_elements(rosetta_resolve_attr(self, "assetClass"), "=", rosetta_resolve_attr(StandardizedScheduleAssetClassEnum, "CREDIT")), _then_fn1, _else_fn1)
    
    percentage =  if_cond_fn(all_elements(rosetta_resolve_attr(self, "assetClass"), "=", rosetta_resolve_attr(StandardizedScheduleAssetClassEnum, "INTEREST_RATES")), _then_fn0, _else_fn0)
    
    
    return percentage

sys.modules[__name__].__class__ = create_module_attr_guardian(sys.modules[__name__].__class__)

File: cdm-sources/cdm/margin/schedule/functions/GetStandardizedScheduleMarginRate.java
package cdm.margin.schedule.functions;

import cdm.margin.schedule.StandardizedScheduleAssetClassEnum;
import com.google.inject.ImplementedBy;
import com.rosetta.model.lib.expression.CardinalityOperator;
import com.rosetta.model.lib.functions.RosettaFunction;
import com.rosetta.model.lib.mapper.MapperS;
import java.math.BigDecimal;

import static com.rosetta.model.lib.expression.ExpressionOperators.*;

@ImplementedBy(GetStandardizedScheduleMarginRate.GetStandardizedScheduleMarginRateDefault.class)
public abstract class GetStandardizedScheduleMarginRate implements RosettaFunction {

	/**
	* @param assetClass 
	* @param durationInYears 
	* @return percentage 
	*/
	public BigDecimal evaluate(StandardizedScheduleAssetClassEnum assetClass, BigDecimal durationInYears) {
		BigDecimal percentage = doEvaluate(assetClass, durationInYears);
		
		return percentage;
	}

	protected abstract BigDecimal doEvaluate(StandardizedScheduleAssetClassEnum assetClass, BigDecimal durationInYears);

	public static class GetStandardizedScheduleMarginRateDefault extends GetStandardizedScheduleMarginRate {
		@Override
		protected BigDecimal doEvaluate(StandardizedScheduleAssetClassEnum assetClass, BigDecimal durationInYears) {
			BigDecimal percentage = null;
			return assignOutput(percentage, assetClass, durationInYears);
		}
		
		protected BigDecimal assignOutput(BigDecimal percentage, StandardizedScheduleAssetClassEnum assetClass, BigDecimal durationInYears) {
			if (areEqual(MapperS.of(assetClass), MapperS.of(StandardizedScheduleAssetClassEnum.INTEREST_RATES), CardinalityOperator.All).getOrDefault(false)) {
				if (lessThanEquals(MapperS.of(durationInYears), MapperS.of(BigDecimal.valueOf(2)), CardinalityOperator.All).getOrDefault(false)) {
					percentage = new BigDecimal("1.0");
				} else if (greaterThan(MapperS.of(durationInYears), MapperS.of(BigDecimal.valueOf(2)), CardinalityOperator.All).and(lessThanEquals(MapperS.of(durationInYears), MapperS.of(BigDecimal.valueOf(5)), CardinalityOperator.All)).getOrDefault(false)) {
					percentage = new BigDecimal("2.0");
				} else if (greaterThan(MapperS.of(durationInYears), MapperS.of(BigDecimal.valueOf(5)), CardinalityOperator.All).getOrDefault(false)) {
					percentage = new BigDecimal("4.0");
				} else {
					percentage = null;
				}
			} else if (areEqual(MapperS.of(assetClass), MapperS.of(StandardizedScheduleAssetClassEnum.CREDIT), CardinalityOperator.All).getOrDefault(false)) {
				if (lessThanEquals(MapperS.of(durationInYears), MapperS.of(BigDecimal.valueOf(2)), CardinalityOperator.All).getOrDefault(false)) {
					percentage = new BigDecimal("2.0");
				} else if (greaterThan(MapperS.of(durationInYears), MapperS.of(BigDecimal.valueOf(2)), CardinalityOperator.All).and(lessThanEquals(MapperS.of(durationInYears), MapperS.of(BigDecimal.valueOf(5)), CardinalityOperator.All)).getOrDefault(false)) {
					percentage = new BigDecimal("5.0");
				} else if (greaterThan(MapperS.of(durationInYears), MapperS.of(BigDecimal.valueOf(5)), CardinalityOperator.All).getOrDefault(false)) {
					percentage = new BigDecimal("10.0");
				} else {
					percentage = null;
				}
			} else if (areEqual(MapperS.of(assetClass), MapperS.of(StandardizedScheduleAssetClassEnum.FOREIGN_EXCHANGE), CardinalityOperator.All).getOrDefault(false)) {
				percentage = new BigDecimal("6.0");
			} else if (areEqual(MapperS.of(assetClass), MapperS.of(StandardizedScheduleAssetClassEnum.EQUITY), CardinalityOperator.All).getOrDefault(false)) {
				percentage = new BigDecimal("15.0");
			} else if (areEqual(MapperS.of(assetClass), MapperS.of(StandardizedScheduleAssetClassEnum.COMMODITY), CardinalityOperator.All).getOrDefault(false)) {
				percentage = new BigDecimal("15.0");
			} else {
				percentage = null;
			}
			
			return percentage;
		}
	}
}
